File: vehicle-tracker-backend/asset-service/src/main/java/com/github/diogocerqueiralima/domain/model/assets/Vehicle.java
package com.github.diogocerqueiralima.domain.model.assets;

import java.time.Instant;
import java.time.LocalDate;
import java.util.Objects;
import java.util.UUID;

/**
 * Represents a vehicle asset in the system.
 * A vehicle has a VIN, plate number, model, manufacturer, and manufacturing date.
 */
public class Vehicle extends Asset {

    private final String vin;
    private final String plate;
    private final String model;
    private final String manufacturer;
    private final LocalDate manufacturingDate;

    public Vehicle(
            UUID id, Instant createdAt, Instant updatedAt, String vin, String plate,
            String model, String manufacturer, LocalDate manufacturingDate
    ) {
        super(id, createdAt, updatedAt);
        this.vin = Objects.requireNonNull(vin, "vin cannot be null");
        this.plate = Objects.requireNonNull(plate, "plate cannot be null");
        this.model = Objects.requireNonNull(model, "model cannot be null");
        this.manufacturer = Objects.requireNonNull(manufacturer, "manufacturer cannot be null");
        this.manufacturingDate = Objects.requireNonNull(manufacturingDate, "manufacturingDate cannot be null");
    }

    public String getVin() {
        return vin;
    }

    public String getPlate() {
        return plate;
    }

    public String getModel() {
        return model;
    }

    public String getManufacturer() {
        return manufacturer;
    }

    public LocalDate getManufacturingDate() {
        return manufacturingDate;
    }

}

File: vehicle-tracker-mobile/app/src/main/java/com/github/diogocerqueiralima/domain/vehicles/model/Vehicle.kt
package com.github.diogocerqueiralima.domain.vehicles.model

import java.time.LocalDate

/**
 * Represents a vehicle asset with its identifying and manufacturing attributes.
 */
data class Vehicle(
    val vin: String,
    val plate: String,
    val model: String,
    val manufacturer: String,
    val manufacturingDate: LocalDate
) {

    init {

        // 1. VIN must follow the standard 17-character format.
        require(vin.isNotBlank()) { "vin must not be blank" }
        require(vin.length == 17) { "vin must be 17 characters long" }

        // 2. Plate, model and manufacturer are required identifying attributes.
        require(plate.isNotBlank()) { "plate must not be blank" }
        require(model.isNotBlank()) { "model must not be blank" }
        require(manufacturer.isNotBlank()) { "manufacturer must not be blank" }

        // 3. A vehicle cannot have been manufactured in the future.
        require(!manufacturingDate.isAfter(LocalDate.now())) { "manufacturingDate must not be in the future" }

    }

    /**
     * Human-readable name composed of the manufacturer and model.
     */
    val displayName = "$manufacturer $model"

}